src/auth/index.ts

```diff
@@ -161,22 +202,45 @@ export class Authenticator {
 
   async authenticate(): Promise<void> {
     // Try to get token from cache using read lock
-    const cachedToken = await this.tryGetCachedToken();
+    const cachedToken = await this.tryGetCachedTokenRecord();
     if (cachedToken) {
-      this.accessToken = cachedToken;
+      this.accessToken = cachedToken.token;
+      this.tokenExpiryTimestampMs = cachedToken.tokenExpiryTimestampMs;
       return;
     }
-
     // No cached token, acquire write lock and authenticate
     await this.acquireWriteLockAndAuthenticate();
```

Comment:
This function already does all the above inside of it. Can we just say `authenticate` == `this.acquireWriteLockAndAuthenticate`?

Reply:
`authenticate` gets read lock to fetch the token then falls back to `acquireWriteLockAndAuthenticate` if looks like cached token needs to be refreshed. It's not equivalent since write lock prevents any other reads and read lock allows multiple locks to be acquired as long as there's no write lock.
Using `acquireWriteLockAndAuthenticate` would mean we're essentially introducing a bottleneck here.

Reply:
Ok I see. It's just the logic duplication confuses me. Can we either
- Only acquire writeLock in `acquireWriteLockAndAuthenticate` after we've validated that no cache entry is present?
- Don't check cache in that function at all and let outer function manage it
At this point the fact that we check cache in multiple places seems confusing to me

Reply:
We have to check the cache twice. Once - under read lock. If it's not cached then we need to acquire write lock and write to it. However, there could be two "threads" at the same stage here and if one acquires a write lock first then it will refresh the token. There's no need for another "thread" to do the same since "thread" 1 has already done so. But we're under the same stage - past the initial check and before the authentication refresh, so after we acquired a write lock we check again that no other "thread" has refreshed the token while we were waiting on a lock.
This efficiency is especially crucial when we have customers that have 600 query/second bursts and if we're doing an auth each time it will grind to a halt.